<launch>

  <arg name="manager" />

  <!-- "camera" should be a user friendly string to uniquely identify the device namespace.
       All topics are pushed down into the "camera" namespace, and it is prepended to tf frame ids. -->
  <arg name="camera"          default="camera" />

  <!-- Type of RealSense camera -->
  <arg name="camera_type"     default="SR300" />

  <!-- "serial_no" or "usb_port_id" should be a unique identifier to connect to a device -->
  <arg name="serial_no"     default="" />
  <arg name="usb_port_id"   default="" /> <!-- USB "Bus#-Port#" -->

  <!-- Setting of stream -->
  <arg name="mode"            default="manual" />
  <arg name="color_width"     default="1920" />
  <arg name="color_height"    default="1080" />
  <arg name="depth_width"     default="640" />
  <arg name="depth_height"    default="480" />
  <arg name="depth_fps"       default="30" />
  <arg name="color_fps"       default="30" />

  <!-- Arguments for remapping all device namespaces -->
  <arg name="depth"            default="depth" />
  <arg name="rgb"              default="rgb" />
  <arg name="ir"               default="ir" />
  <arg name="depth_registered" default="depth_registered" />

  <!-- Size of message queue for synchronizing subscribed topics in depth_image_proc package.
       Increase this value for higher resolution and/or frame rates.
       If the depth_registered pointcloud topic generated by the depth_image_proc package does not contain data,
       increase this value further. -->
  <arg name="queue_size"       default="100" />
  <param name="$(arg camera)/points_xyzrgb_sw_registered/queue_size" type="int" value="$(arg queue_size)" />

  <!-- Optionally suppress loading the driver nodelet. Useful while playing back recorded raw data from a bag.-->
  <arg name="load_driver"   default="true" />
  <arg name="publish_tf"    default="true" />

  <!-- Processing Modules -->
  <arg name="rgb_processing"                  default="true"/>
  <arg name="ir_processing"                   default="true"/>
  <arg name="depth_processing"                default="true"/>
  <arg name="depth_registered_processing"     default="true"/>
  <arg name="disparity_processing"            default="true"/>
  <arg name="disparity_registered_processing" default="true"/>
  <arg name="hw_registered_processing"        default="false" />
  <arg name="sw_registered_processing"        default="true" />

  <!-- Disable respawn topics by default -->
  <arg name="respawn" default="false" />

  <!-- Push down all topics/nodelets into "camera" namespace -->
  <group ns="$(arg camera)">

    <!-- Load driver -->
    <include file="$(find realsense_camera)/launch/includes/nodelet_rgbd.launch.xml">
      <arg name="manager"                   value="$(arg manager)" />
      <arg name="camera"                    value="$(arg camera)" />
      <arg name="camera_type"               value="$(arg camera_type)" />
      <arg name="serial_no"                 value="$(arg serial_no)" />
      <arg name="usb_port_id"               value="$(arg usb_port_id)" />
      <arg name="depth"                     value="$(arg depth)" />
      <arg name="rgb"                       value="$(arg rgb)" />
      <arg name="ir"                        value="$(arg ir)" />
      <arg name="depth_processing"          value="$(arg depth_processing)" />
      <arg name="rgb_processing"            value="$(arg rgb_processing)" />
      <arg name="ir_processing"             value="$(arg ir_processing)" />
      <arg name="publish_tf"                value="$(arg publish_tf)" />
      <arg name="mode"                      value="$(arg mode)" />
      <arg name="depth_width"               value="$(arg depth_width)" />
      <arg name="depth_height"              value="$(arg depth_height)" />
      <arg name="color_width"               value="$(arg color_width)" />
      <arg name="color_height"              value="$(arg color_height)" />
      <arg name="depth_fps"                 value="$(arg depth_fps)" />
      <arg name="color_fps"                 value="$(arg color_fps)" />
    </include>

    <!-- Load standard constellation of processing nodelets -->
    <include file="$(find rgbd_launch)/launch/includes/processing.launch.xml">
      <arg name="manager"                         value="$(arg manager)" />
      <arg name="rgb"                             value="$(arg rgb)" />
      <arg name="ir"                              value="$(arg ir)" />
      <arg name="depth"                           value="$(arg depth)" />
      <arg name="depth_registered"                value="$(arg depth_registered)" />
      <arg name="respawn"                         value="$(arg respawn)" />
      <arg name="rgb_processing"                  value="$(arg rgb_processing)" />
      <arg name="ir_processing"                   value="$(arg ir_processing)" />
      <arg name="depth_processing"                value="$(arg depth_processing)" />
      <arg name="depth_registered_processing"     value="$(arg depth_registered_processing)" />
      <arg name="disparity_processing"            value="$(arg disparity_processing)" />
      <arg name="disparity_registered_processing" value="$(arg disparity_registered_processing)" />
      <arg name="hw_registered_processing"        value="$(arg hw_registered_processing)" />
      <arg name="sw_registered_processing"        value="$(arg sw_registered_processing)" />
    </include>
    
    <!-- Load capture realsense nodelet -->
    <node pkg="nodelet" type="nodelet" name="robot_vision_capture_realsense" 
          args="load T2_robot_vision/robot_vision_capture_realsense $(arg manager)">
      <remap from="camera/rgb/image_rect_color" to="rgb/image_rect_color" />
      <remap from="camera/depth/image_rect_raw" to="depth/image_rect_raw" />
      <remap from="camera/depth/points" to="depth/points" />
      <remap from="camera/depth/camera_info" to="depth/camera_info" />
      <remap from="camera/fusion/image_c2d" to="fusion/image_c2d" />
      <remap from="camera/fusion/image_d2c" to="fusion/image_d2c" />
      <remap from="camera/driver/force_power" to="driver/force_power" />
    </node>

  </group> <!-- camera -->

</launch>
